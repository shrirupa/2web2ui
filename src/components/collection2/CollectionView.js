import React, { Component } from 'react';
import CollectionViewPropTypes from './CollectionView.propTypes';
import { Loading } from 'src/components';

const PassThroughWrapper = (props) => props.children;
const NullComponent = () => null;

const defaultFormatRow = (row) => Object.keys(row).map((fieldName) => row[fieldName]);

class CollectionView extends Component {
  render() {
    const {
      rows,
      formatRow = defaultFormatRow,
      rowComponent: RowComponent,
      rowKeyName = 'id',
      headerComponent: HeaderComponent = NullComponent,
      outerWrapper: OuterWrapper = PassThroughWrapper,
      bodyWrapper: BodyWrapper = PassThroughWrapper
    } = this.props;

    if (!Array.isArray(rows)) {
      return <Loading />;
    }

    if (!rows.length) {
      return null; // nothing to display msg?
    }

    return <OuterWrapper>
      <HeaderComponent />
      <BodyWrapper>
        {rows.map((row, i) => <RowComponent key={`${row[rowKeyName] || 'row'}-${i}`} rowData={formatRow(row)} />)}
      </BodyWrapper>
    </OuterWrapper>;
  }
}

CollectionView.propTypes = CollectionViewPropTypes;

export default CollectionView;
