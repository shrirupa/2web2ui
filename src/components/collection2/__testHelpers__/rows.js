export function mkRow(i = 0) {
  return { id: i, name: `row-name-${Math.floor(i / 5)}-${i}` };
}

export function mkRows(n) {
  const rows = [];
  for (let i = 0; i < n; i++) {
    rows.push(mkRow(i));
  }
  return rows;
}
