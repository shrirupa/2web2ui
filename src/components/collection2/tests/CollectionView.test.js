import React from 'react';
import { shallow } from 'enzyme';
import CollectionView from '../CollectionView';
import { mkRows } from '../__testHelpers__/rows';

describe('Component: CollectionView', () => {
  let props;

  beforeEach(() => {
    props = {
      rows: [],
      rowComponent: () => <h1>row</h1>,
      headerComponent: () => <h1>header</h1>
    };
  });

  it('should render null if there are no rows', () => {
    const wrapper = shallow(<CollectionView {...props} />);
    expect(wrapper.equals(null)).toEqual(true);
  });

  it('should render correctly with basic props', () => {
    const wrapper = shallow(<CollectionView {...props} rows={mkRows(15)} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with custom wrappers', () => {
    props.outerWrapper = function CustomOuterWrapper(props) { return props.children; };
    props.bodyWrapper = function CustomBodyWrapper(props) { return props.children; };
    expect(shallow(<CollectionView {...props} rows={mkRows(3)} />)).toMatchSnapshot();
  });
});
