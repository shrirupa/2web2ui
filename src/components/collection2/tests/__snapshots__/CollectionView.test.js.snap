// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CollectionView should render correctly with basic props 1`] = `
<PassThroughWrapper>
  <headerComponent />
  <PassThroughWrapper>
    <rowComponent
      id={0}
      key="row-0"
      name="row-name--0"
    />
    <rowComponent
      id={1}
      key="1-1"
      name="row-name--1"
    />
    <rowComponent
      id={2}
      key="2-2"
      name="row-name--2"
    />
    <rowComponent
      id={3}
      key="3-3"
      name="row-name--3"
    />
    <rowComponent
      id={4}
      key="4-4"
      name="row-name--4"
    />
    <rowComponent
      id={5}
      key="5-5"
      name="row-name--5"
    />
    <rowComponent
      id={6}
      key="6-6"
      name="row-name--6"
    />
    <rowComponent
      id={7}
      key="7-7"
      name="row-name--7"
    />
    <rowComponent
      id={8}
      key="8-8"
      name="row-name--8"
    />
    <rowComponent
      id={9}
      key="9-9"
      name="row-name--9"
    />
    <rowComponent
      id={10}
      key="10-10"
      name="row-name--10"
    />
    <rowComponent
      id={11}
      key="11-11"
      name="row-name--11"
    />
    <rowComponent
      id={12}
      key="12-12"
      name="row-name--12"
    />
    <rowComponent
      id={13}
      key="13-13"
      name="row-name--13"
    />
    <rowComponent
      id={14}
      key="14-14"
      name="row-name--14"
    />
  </PassThroughWrapper>
</PassThroughWrapper>
`;

exports[`Component: CollectionView should render with custom wrappers 1`] = `
<CustomOuterWrapper>
  <headerComponent />
  <CustomBodyWrapper>
    <rowComponent
      id={0}
      key="row-0"
      name="row-name--0"
    />
    <rowComponent
      id={1}
      key="1-1"
      name="row-name--1"
    />
    <rowComponent
      id={2}
      key="2-2"
      name="row-name--2"
    />
  </CustomBodyWrapper>
</CustomOuterWrapper>
`;
